### PR TITLE
Uart interrupt

### PR DESCRIPTION
**Pull request by ClarkPoon**
implemented fifo for storing characters and interrupts for uart receiving. Also have support for receiving, storing, and getting UART commands for cli.

### CODE REVIEW

File: BeVolt/App/src/main.c [hunk 1]
```diff
@@ -593,7 +593,7 @@ int SPITestmain(){
 // Debug UART Test
 #include "UART.h"
 int main(){
-	UART3_Init(115200);
+	UART1_Init(115200);
```
Comment (chaseblock) [inline BeVolt/App/src/main.c]: According to @SijWoo, the new master board is using UART3.

File: BeVolt/App/src/main.c [hunk 2]
```diff
@@ -762,4 +762,164 @@ int main() {
...
+
+#ifdef UART_INTERRUPT2
+    /* Includes ---------------------------------------------------*/
+    //#include "stm32f2xx.h"
+		#include "uart.h"
+		
+		uint8_t RxData;//data received flag
+				
+    /* Private function prototypes --------------------------------*/
+    void USART1_Config(void);
```
Comment (chaseblock) [inline BeVolt/App/src/main.c]: UART3?

File: BeVolt/App/src/main.c [hunk 3]
```diff
@@ -762,4 +762,164 @@ int main() {
...
+
+    /* Private functions ------------------------------------------*/
+		
+		/**
+		 * @brief This function handles USART3 global interrupt request.
+     * @param None
+     * @retval None
+     */
+    void USART1_IRQHandler(void){
```
Comment (chaseblock) [inline BeVolt/App/src/main.c]: UART3?

File: BeVolt/App/src/main.c [hunk 4]
```diff
@@ -762,4 +762,164 @@ int main() {
...
+			if(USART_GetITStatus(USART1, USART_IT_RXNE) != RESET){
```
Comment (chaseblock) [inline BeVolt/App/src/main.c]: USART3

File: BeVolt/App/src/main.c [hunk 5]
```diff
@@ -762,4 +762,164 @@ int main() {
...
+     /* Read one byte from the receive data register */
+          RxData = USART_ReceiveData(USART1);
```
Comment (chaseblock) [inline BeVolt/App/src/main.c]: Etc.

File: BeVolt/App/src/main.c [hunk 6]
```diff
@@ -762,4 +762,164 @@ int main() {
...
+     /****************************************************************
+      -Step3-
+      Implement the peripheral interrupt handler (PPP_IRQHandler)
+      in stm32f2xx_it.c file.
+     ****************************************************************/
+      /* Refer to next section "stm32f2xx_it.c" */
+
+      while (1)
+      {
+				//do something important
```
Comment (chaseblock) [inline BeVolt/App/src/main.c]: Come of these comments could be a little more specific.

File: BeVolt/App/src/main.c [hunk 7]
```diff
@@ -762,4 +762,164 @@ int main() {
...
+						fifo[head] = 0x0a;//put line feed in
+						head = ((head + 1) % size);
+					}
+					RxData = 0;
+				}
+				if (head != tail) {
+					printf("%c", fifo[tail]);
+					tail = (tail +1) % size;
+				}else{
+					//printf(".");
```
Comment (chaseblock) [inline BeVolt/App/src/main.c]: Dead code.

File: BeVolt/App/src/main.c [hunk 8]
```diff
@@ -762,4 +762,164 @@ int main() {
...
+				}
+      }
+     }
+	 
+     /**
+      * @brief Configures the USART3 Peripheral.
```
Comment (chaseblock) [inline BeVolt/App/src/main.c]: Doesn't seem to be configuring USART3, even though it should

File: BeVolt/Periphs/inc/UART.h [hunk 9]
```diff
@@ -53,4 +58,34 @@ void UART3_Write(char *txBuf, uint32_t txSize);
...
+/**
+* @brief This function handles USART1 global interrupt request.
+* @param None
+* @retval None
+*/
+void USART1_IRQHandler(void);
+
+//use when fifo is storing commands received over uart
+//returns if fifo has a command in it by checking for carriage return
+bool hasCommand(Fifo *fifo);
```
Comment (chaseblock) [inline BeVolt/Periphs/inc/UART.h]:
```suggestion
bool USART_HasCommand(Fifo *fifo);
```

File: BeVolt/Periphs/inc/UART.h [hunk 10]
```diff
@@ -53,4 +58,34 @@ void UART3_Write(char *txBuf, uint32_t txSize);
...
+
+//for using fifo to store commands received from UART
+//fails if multiple commands are stored in queue
+//preconditions: fifo has a command stored in it, buffer is large enough to hold command
+//parameters: fifo with command stored in it, buffer to receive stored command
+//fills buffer with received command
+void getCommand(Fifo *fifo, char buffer[]);
```
Comment (chaseblock) [inline BeVolt/Periphs/inc/UART.h]:
```suggestion
void USART_GetCommand(Fifo *fifo, char buffer[]);
```

File: BeVolt/Periphs/inc/UART.h [hunk 11]
```diff
@@ -53,4 +58,34 @@ void UART3_Write(char *txBuf, uint32_t txSize);
...
+
+//checks if new data has been received from UART interrupts
+//if so, it echoes the new character to Putty and adds the character to the passed fifo
+//backspace is supported
+void checkUARTandEcho(Fifo *uartFifo);
```
Comment (chaseblock) [inline BeVolt/Periphs/inc/UART.h]:
```suggestion
void USART_CheckUARTandEcho(Fifo *uartFifo);
```

File: BeVolt/App/src/main.c (hunk 1 above)
Comment (SijWoo) [reply BeVolt/App/src/main.c]: After this semester, I'm officially deprecating the 1st version board so we won't ever need to use UART1 for the STM version

Comment (chaseblock) [reply BeVolt/App/src/main.c]: Sounds good to me. We should make sure that we have a second copy (at least) of the master board before deprecation.

**Pull request merged**